### PR TITLE
feat: add code source references and related page tagging; emits Rela…

### PR DESCRIPTION
**Pull request by notowen333**
# Tag-driven Related Pages

**note** this change does not change any human facing pages. It just adds related pages and source reference metadata for agents and headless browsers. As a follow up we /may/ add tags at the bottom of pages that group other related pages.

`sourceLinks` infrastructure (schema, renderer for `## Implementation` blocks) is wired but no values are set — pending the upcoming monorepo migration. Re-adoption is purely a frontmatter exercise.

## Surfaces

| Surface | Where | Cap | Score floor | Audience |
|---|---|---|---|---|
| **`## Related pages` block** | appended to body in `/<slug>/index.md` and aggregated in `/llms-full.txt` | 10 | none | LLMs / `llms.txt` consumers |
| **JSON-LD `relatedLink`** | inside the page's `TechArticle` graph node | 10 | none | HTML-only crawlers |

## Algorithm

Score is rarity-weighted Jaccard:

```
score =  Σ (rarity-weight of tags in BOTH pages)
        ─────────────────────────────────────────
         Σ (rarity-weight of tags in EITHER page)
```

Each tag's rarity-weight is `1 - (pages_with_tag / total_tagged_pages)`. Common tags (e.g. `aws`, used by 14 pages) contribute near 0; rare tags (e.g. `agentcore`, used by 4) contribute near 1.

A 1-tag match on a broad tag ranks below any 2-tag specific match, so coincidental connections sink naturally. A tag that bloats over time loses weight automatically — no manual audit cycle needed to react. Ties break alphabetically by title (deterministic).

## Tag registry

[`src/config/tags.yml`](src/config/tags.yml) — 32 tags grouped by axis (topics / capabilities / lifecycle), validated by Zod at build time. The YAML header documents the rule:

> A tag means **"this page teaches the named thing."** Not "mentions" or "has a section labeled with it."

…with a four-trap checklist drawn from concrete failure modes encountered while tagging the corpus.

## Coverage

112 of 123 user-guide pages tagged. 11 are deliberately untagged (umbrella pages, policy docs, pages with no good cross-section bridge).

## Concrete output for `safety-security/guardrails`

Source frontmatter:

```yaml
title: Guardrails
tags: [safety, bedrock, aws]
```

### `/docs/user-guide/safety-security/guardrails/index.md` and the same block inside `/llms-full.txt`

Top 10, ranked by score, no floor:

```markdown
## Related pages

- [Amazon Bedrock](/docs/user-guide/concepts/model-providers/amazon-bedrock/index.md) (3 shared tags)
- [Amazon Nova](/docs/user-guide/concepts/model-providers/amazon-nova/index.md) (2 shared tags)
- [Nova Sonic](/docs/user-guide/concepts/bidirectional-streaming/models/nova_sonic/index.md) (2 shared tags)
- [Deploying Strands Agents to Amazon Bedrock AgentCore Runtime](/docs/user-guide/deploy/deploy_to_bedrock_agentcore/index.md) (2 shared tags)
- [Python Deployment to Amazon Bedrock AgentCore Runtime](/docs/user-guide/deploy/deploy_to_bedrock_agentcore/python/index.md) (2 shared tags)
- [TypeScript Deployment to Amazon Bedrock AgentCore Runtime](/docs/user-guide/deploy/deploy_to_bedrock_agentcore/typescript/index.md) (2 shared tags)
- [AgentCore Evaluation Dashboard Configuration](/docs/user-guide/evals-sdk/how-to/agentcore_evaluation_dashboard/index.md) (2 shared tags)
- [PII Redaction](/docs/user-guide/safety-security/pii-redaction/index.md) (2 shared tags)
- [Harmfulness Evaluator](/docs/user-guide/evals-sdk/evaluators/harmfulness_evaluator/index.md) (1 shared tag)
- [Refusal Evaluator](/docs/user-guide/evals-sdk/evaluators/refusal_evaluator/index.md) (1 shared tag)
```

Links are emitted as `index.md` siblings so an LLM following them stays on the markdown surface.

### `<head>` JSON-LD `TechArticle.relatedLink`

Same 10 entries, same ordering, but as canonical HTML URLs:

```json
{
  "@type": "TechArticle",
  "headline": "Guardrails",
  "keywords": "safety, bedrock, aws",
  "relatedLink": [
    "https://strandsagents.com/docs/user-guide/concepts/model-providers/amazon-bedrock/",
    "https://strandsagents.com/docs/user-guide/concepts/model-providers/amazon-nova/",
    /* ... */
  ]
}
```

### HTML "See also" pills

Top 6 from the same ranking, filtered to score ≥ 0.4. On Guardrails this lands as 6 pills; on a thin-tag page (e.g. `retry-strategies` whose best score is 0.34) the strip is empty, by design.

## Safety nets

- **Zod validation** on tag registry — unknown tags fail the build at the page level.
- **Title-uniqueness lint** in `test/content-collection.test.ts` — fails build on cross-section title collisions (the "two pages named Hooks" failure mode that produces ambiguous pills).
- **14 algorithm tests** in `test/related-docs.test.ts` covering scoring, tie-breaking, score floor, edge cases. 311 tests total in the repo.

## Verification

- `npm run build` — clean, no broken links across 515 pages
- `npm run typecheck` — clean
- `npm test` — 311 pass

## Type of Change

- New feature

## Checklist

- [x] I have read the CONTRIBUTING document
- [x] My changes follow the project's documentation style
- [x] I have tested the documentation locally using `npm run dev`
- [x] Links in the documentation are valid and working

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/components/overrides/Head.astro [hunk 1]
```diff
@@ -60,14 +62,48 @@ const breadcrumbItems = pathSegments.map((segment, i) => {
...
 
+// TechArticle node for docs pages with tags or sourceLinks.
+// Reaches HTML-only crawlers that don't follow the llms.txt convention.
+const entry = route.entry
+// Non-docs routes (blog authors/tags pages, etc.) flow through Starlight too
+// but their synthetic entry.data has no `tags` field — guard with ??.
+const tags = entry.data.tags ?? []
+const sourceLinks = entry.data.sourceLinks ?? []
+let techArticle: Record<string, unknown> | null = null
+if (tags.length > 0 || sourceLinks.length > 0) {
```
Comment (zastrowm) [inline src/components/overrides/Head.astro:80]: Anything we could link to here for context (if we wanted to know what fields are available)

File: src/util/related-docs.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  allDocs: readonly CollectionEntry<'docs'>[],
+): RelatedLink[] {
+  if (!isUserGuide(current)) return []
+  // `?? []` guards non-docs Starlight routes (blog, StarlightPage) that flow
+  // through the same components but whose synthetic entry.data lacks `tags`.
+  const currentTags = current.data.tags ?? []
+  if (currentTags.length === 0) return []
+  const tagSet = new Set<string>(currentTags)
+
+  return allDocs
```
Comment (zastrowm) [inline src/util/related-docs.ts:112]: This would be a good place to have a summary comment of what's going on below

File: src/content/docs/user-guide/concepts/agents/retry-strategies.mdx [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ---
 title: Retry Strategies
+tags: [retry]
```
Comment (zastrowm) [inline src/content/docs/user-guide/concepts/agents/retry-strategies.mdx]: Does retry really have enough to warrant a tag?

File: src/content/docs/user-guide/concepts/agents/state.mdx [hunk 4]
```diff
@@ -2,6 +2,7 @@
 title: State Management
 sidebar:
   label: "State"
+tags: [state]
```
Comment (zastrowm) [inline src/content/docs/user-guide/concepts/agents/state.mdx]: Snapshots would be good for state and/or session-management

File: src/content/docs/user-guide/concepts/bidirectional-streaming/events.mdx [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ---
 title: Events
 experimental: true
+tags: [streaming]
```
Comment (zastrowm) [inline src/content/docs/user-guide/concepts/bidirectional-streaming/events.mdx]: Why streaming here? I feel like all of bidi is getting streaming... which maybe, but if you're looking at a single agent streaming, all of this is unrelated IMHO

File: src/content/docs/user-guide/concepts/tools/vended-tools.mdx [hunk 6]
```diff
@@ -4,6 +4,7 @@ description: "Pre-built tools included in the TypeScript SDK for common agent ta
 sidebar:
   label: "Vended Tools"
 languages: [typescript]
+tags: [tools, typescript]
```
Comment (zastrowm) [inline src/content/docs/user-guide/concepts/tools/vended-tools.mdx]: Not sure if language tag is going to get us anything

File: src/content/docs/user-guide/deploy/operating-agents-in-production.mdx [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ---
 title: Operating Agents in Production
 description: "Deploy AI agents to production on AWS Lambda, EKS, Fargate, EC2, or Bedrock AgentCore. Monitoring, scaling, and operational best practices."
+tags: [deployment, production, observability]
```
Comment (zastrowm) [inline src/content/docs/user-guide/deploy/operating-agents-in-production.mdx]: Is deployment=production?  If so can we ditch one?

File: src/config/tags.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+# is validated against this list by the docs schema (src/content.config.ts) —
+# unknown tags fail the build.
+#
+# Scoring: `more shared tags = more related`. Aim for 2–4 tags per page drawn
+# from different axes below so pages naturally connect along topic, capability,
+# lifecycle, and audience dimensions.
+#
+# Adding a tag = PR editing this file in the same change that first uses it.
+
+# Topics / domains — specific products, protocols, data types
```
Comment (zastrowm) [inline src/config/tags.yml:31]: What's the counts for each of these?  Any that are 1s?

Maybe have a test that outputs it for awareness; anything with 1 or 2 might be worth ditching IMO

File: src/content/docs/user-guide/concepts/agents/retry-strategies.mdx (hunk 3 above)
Comment (notowen333) [reply src/content/docs/user-guide/concepts/agents/retry-strategies.mdx]: That's fair...

maybe `event-loop`

File: src/content/docs/user-guide/concepts/bidirectional-streaming/events.mdx (hunk 5 above)
Comment (notowen333) [reply src/content/docs/user-guide/concepts/bidirectional-streaming/events.mdx]: Could go for `bidi-streaming` and just link all of the articles in that category...

File: src/content/docs/user-guide/deploy/operating-agents-in-production.mdx (hunk 7 above)
Comment (notowen333) [reply src/content/docs/user-guide/deploy/operating-agents-in-production.mdx]: I can see a world for both... might be that the first pass of examples is seeing high overlap.

like an article about hello-worlding strands on AC runtime would only get `deployment`. An article about comprehensive tracing only gets `production`

File: src/config/tags.yml (hunk 8 above)
Comment (notowen333) [reply src/config/tags.yml:31]: I think we'd want to take this question on after we do a follow up sweep with more tags. 

What do you mean by ditching? We could for example have a minimum tag overlap count for what we display in the page?

File: src/content/docs/user-guide/concepts/tools/vended-tools.mdx (hunk 6 above)
Comment (notowen333) [reply src/content/docs/user-guide/concepts/tools/vended-tools.mdx]: Sure happy to remove this one

File: src/content/docs/user-guide/deploy/operating-agents-in-production.mdx (hunk 7 above)
Comment (zastrowm) [reply src/content/docs/user-guide/deploy/operating-agents-in-production.mdx]: I guess I'm a little concerned that we're adding two tags here for no purpose and its' just repetitive/noise

File: src/config/tags.yml (hunk 8 above)
Comment (zastrowm) [reply src/config/tags.yml:31]: I mean, if we have a tag that has 1 page tagged, we should just remove the tag.  I don't want to get into the habit of having 20 "1 page" tags that are just noise

Comment (notowen333) [reply src/config/tags.yml:31]: Got it I understand what you mean

Let's take this in a follow up where we do a more comprehensive addition of tags to pages if that's OK with you?

File: src/content/docs/user-guide/deploy/operating-agents-in-production.mdx (hunk 7 above)
Comment (notowen333) [reply src/content/docs/user-guide/deploy/operating-agents-in-production.mdx]: Ok I can remove deployment for now while there is significant overlap

File: src/config/tags.yml (hunk 8 above)
Comment (notowen333) [reply src/config/tags.yml:31]: adding all the tags in the first pass

**Review by zastrowm — APPROVED** [review]
